File: cpp/Problem004_LargestPalindromeProduct.cpp
#include <iostream>
#define MIN 900
#define MAX 1000
using std::cout;

// if reverse of the number == number it is a palindrome
bool isPalindrome(int num)
{
    int reverse = 0, remainder, temp;
    temp = num; // original number is stored in temp variable
    while( temp != 0)
    {
        remainder = temp % 10;
        reverse = reverse * 10 + remainder;
        temp = temp / 10;
    }

    return (num == reverse);
}

int main()
{
    int max = 0;

    for (int i=MIN; i<MAX; i++)
        for (int j=MIN; j<=std::min(i, MAX); j++) // 90*91 == 91*90  so we block it out by doing  i > j
            if (isPalindrome(i*j)) 
                if (max < i*j)
                    max = i*j;

    cout << max << "\n";
}

File: _try_out_langs_/rust/Problem004_LargestPalindromeProduct.rs
fn main() {
    let mut max = 0;
    for a in 900..1000 {
        for b in 900..1000 {
            let v = a*b;
            if is_palindrome(v) && v > max {
                max = v;
            }
        }
    }
    println!("{}", max);
}

fn is_palindrome(v: i32) -> bool {
    let mut reverse = 0;
    let mut tmp = v;

    while tmp > 0 {
        reverse *= 10;
        reverse += tmp % 10;
        tmp /= 10;
    }

    // println!("{} == {}", v, reverse);
    return v == reverse;
}